unit_tests/ex_suffix_expander.cc

```diff
@@ -9,13 +9,177 @@
+  static constexpr TextView GB = "GB";
+  static constexpr TextView TB = "TB";
+  static constexpr TextView PB = "PB";
+
+  static constexpr TextView baseUnit = B;
+
+  static uintmax_t convertToBase(TextView const unit) {
+    if (unit == B) {
+      return static_cast<uintmax_t>(1);
+    } else if (unit == KB) {
+      return static_cast<uintmax_t>(1) << 10;
+    } else if (unit == MB) {
+      return static_cast<uintmax_t>(1) << 20;
+    } else if (unit == GB) {
+      return static_cast<uintmax_t>(1) << 30;
+    } else if (unit == TB) {
+      return static_cast<uintmax_t>(1) << 40;
+    } else if (unit == PB) {
+      return static_cast<uintmax_t>(1) << 50;
+    } else {
+      // TODO: error handling
+      return static_cast<uintmax_t>(0);
+    }
+  }
+
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "b") || iequal(unit, "byte")) {
+      return B;
+    } else if (iequal(unit, "k") || iequal(unit, "kb")) {
+      return KB;
+    } else if (iequal(unit, "m") || iequal(unit, "mb")) {
+      return MB;
+    } else if (iequal(unit, "g") || iequal(unit, "gb")) {
+      return GB;
+    } else if (iequal(unit, "t") || iequal(unit, "tb")) {
+      return TB;
+    } else if (iequal(unit, "p") || iequal(unit, "pb")) {
+      return PB;
+    } else {
+      // TODO: error handling
+      return B;
+    }
+  }
+};
+
+struct Duration {
+  static constexpr TextView SECOND = "second";
+  static constexpr TextView MINUTE = "minute";
+  static constexpr TextView HOUR   = "hour";
+  static constexpr TextView DAY    = "day";
+  static constexpr TextView WEEK   = "week";
+
+  static constexpr TextView baseUnit = "second";
+
+  static uintmax_t convertToBase(TextView const unit) {
+    if (unit == SECOND) {
+      return static_cast<uintmax_t>(1);
+    } else if (unit == MINUTE) {
+      return static_cast<uintmax_t>(60);
+    } else if (unit == HOUR) {
+      return static_cast<uintmax_t>(3'600);
+    } else if (unit == DAY) {
+      return static_cast<uintmax_t>(86'400);
+    } else if (unit == WEEK) {
+      return static_cast<uintmax_t>(604'800);
+    } else {
+      // TODO: error handling
+      return 0;
+    }
+  }
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "s") || iequal(unit, "sec") || iequal(unit, "second")) {
+      return SECOND;
+    } else if (iequal(unit, "m") || iequal(unit, "min") || iequal(unit, "minute")) {
+      return MINUTE;
+    } else if (iequal(unit, "h") || iequal(unit, "hour")) {
+      return HOUR;
+    } else if (iequal(unit, "d") || iequal(unit, "day")) {
+      return DAY;
+    } else if (iequal(unit, "w") || iequal(unit, "week")) {
+      return WEEK;
+    } else {
+      // TODO: error handling
+      return SECOND;
+    }
+  }
+};
+
+} // namespace metric
+
+// (potentially) modifies the pass-in view. returns the unit
+template <typename Metric> inline TextView rExtractUnit(TextView &src) noexcept {
+  src.rtrim_if(&isspace);
+  if (isdigit(src.back())) {
```

Comment:
Should probably check for `empty` otherwise `back` is unreliable.

Reply:
this function is called at the beginning of a while loop conditioned by `while (text.rtrim_if(&isspace))`, so the fact that the iteration is entered means L150 is safe; but at the same time it also makes L149 unnecessary... 

Do you think I should keep the function robust on its own by keeping L149 and add a emptiness check for L150; or remove them both?

Reply:
I think I'd remove the `rtrim_if` here and note this requires a non-empty non-space terminated input. It's purely an internal function, there's no need for it to be generally robust.

Reply:
ack